src/mjolnir/landmark_database_builder.cc

```diff
@@ -220,6 +257,90 @@ void LandmarkDatabase::release_prepared_stmt() {
   LOG_INFO("released prepared statements");
 }
 
+// Task 2: parse landmarks
+// anonymous namespace?
+struct landmark_callback : public OSMPBF::Callback {
+public:
+  landmark_callback(std::vector<Landmark>& landmarks) : landmarks_(landmarks) {
+  }
+  virtual ~landmark_callback() {
+  }
+
+  virtual void
+  node_callback(const uint64_t /*osmid*/, double lng, double lat, const OSMPBF::Tags& tags) override {
+    Landmark landmark;
+
+    for (const auto& tag : tags) {
+      if (tag.first == "amenity") {
+        // if amenity is empty will return LandmarkType::NA
+        landmark.type = string_to_landmark_type(tag.second);
+      }
+      if (tag.first == "name" && !tag.second.empty()) {
+        landmark.name = tag.second;
+      }
+      landmark.lng = lng;
+      landmark.lat = lat;
+    }
+
+    landmarks_.push_back(std::move(landmark));
+  }
+
+  virtual void changeset_callback(const uint64_t changeset_id) override {
+  }
+
+  virtual void way_callback(const uint64_t /*osmid*/,
+                            const OSMPBF::Tags& /*tags*/,
+                            const std::vector<uint64_t>& /*nodes*/) override {
+    LOG_WARN("way callback shouldn't be called!");
+  }
+
+  virtual void relation_callback(const uint64_t /*osmid*/,
+                                 const OSMPBF::Tags& /*tags*/,
+                                 const std::vector<OSMPBF::Member>& /*members*/) override {
+    LOG_WARN("relation callback shouldn't be called!");
+  }
+
+  std::vector<Landmark>& landmarks_;
+};
+
+bool BuildLandmarkFromPBF(const std::vector<std::string>& input_files) {
+  // parse nodes in pbf to get landmarks
+  std::vector<Landmark> landmarks{};
+
+  landmark_callback callback(landmarks);
+
+  LOG_INFO("Parsing files...");
+  // hold open all the files so that if something else (like diff application)
+  // needs to mess with them we wont have troubles with inodes changing underneath us
+  std::list<std::ifstream> file_handles;
+  for (const auto& input_file : input_files) {
+    file_handles.emplace_back(input_file, std::ios::binary);
+    if (!file_handles.back().is_open()) {
+      throw std::runtime_error("Unable to open: " + input_file);
+    }
+  }
+
+  LOG_INFO("Parsing nodes...");
+  for (auto& file_handle : file_handles) {
+    OSMPBF::Parser::parse(file_handle,
+                          static_cast<OSMPBF::Interest>(OSMPBF::Interest::NODES |
+                                                        OSMPBF::Interest::CHANGESETS),
+                          callback);
+  }
+
+  // store landmarks in database
+  LandmarkDatabase db(landmark_db, false);
+
+  for (const auto& landmark : landmarks) {
+    if (!db.insert_landmark(landmark)) {
```

Comment:
do this inside the callback for each one you parse. there is no sense in allocating a giant vector when you can just push a landmark into the db immediately as soon as its parsed. this means the parser needs the landmarkdatabase as a member though of course

Reply:
good idea. I separated parser and insert because I wanted to test if parser is correct. now I know it is so can put insert into callback directly.